src/icrar/leap-accelerate/cuda/MetaDataCuda.cu

```diff
@@ -0,0 +1,192 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+ * MA 02111 - 1307  USA
+ */
+
+#include "MetaDataCuda.h"
+#include <icrar/leap-accelerate/math/math.h>
+#include <icrar/leap-accelerate/math/casacore_helper.h>
+
+namespace icrar
+{
+namespace cuda
+{
+    bool Constants::operator==(const Constants& rhs) const
+    {
+        return nantennas == rhs.nantennas
+        && channels == rhs.channels
+        && num_pols == rhs.num_pols
+        && stations == rhs.stations
+        && rows == rhs.rows
+        && freq_start_hz == rhs.freq_start_hz
+        && freq_inc_hz == rhs.freq_inc_hz
+        && channel_wavelength == rhs.channel_wavelength
+        && phase_centre_ra_rad == rhs.phase_centre_ra_rad
+        && phase_centre_dec_rad == rhs.phase_centre_dec_rad
+        && dlm_ra == rhs.dlm_ra
+        && dlm_dec == rhs.dlm_dec;
+    }
+
+    MetaDataCudaHost::MetaDataCudaHost(const MetaData& metadata)
+    {
+        m_constants.nantennas = metadata.nantennas;
+        m_constants.channels = metadata.channels;
+        m_constants.num_pols = metadata.num_pols;
+        m_constants.stations = metadata.stations;
+        m_constants.rows = metadata.rows;
+        m_constants.freq_start_hz = metadata.freq_start_hz;
+        m_constants.freq_inc_hz = metadata.freq_inc_hz;
+        m_constants.phase_centre_ra_rad = metadata.phase_centre_ra_rad;
+        m_constants.phase_centre_dec_rad = metadata.phase_centre_dec_rad;
+        m_constants.dlm_ra = metadata.dlm_ra;
+        m_constants.dlm_dec = metadata.dlm_dec;
+
+        init = metadata.init;
+        oldUVW = metadata.oldUVW;
+
+        avg_data = ConvertMatrix(metadata.avg_data);
+        
+        if(metadata.dd.is_initialized())
+        {
+            dd = ConvertMatrix3x3(metadata.dd.value());
+        }
+        else
+        {
+            throw std::runtime_error("dd is required");
+        }
+
+        A = ConvertMatrix(metadata.A);
+        I = ConvertMatrix<int>(metadata.I);
+        Ad = ConvertMatrix(metadata.Ad);
+
+        A1 = ConvertMatrix(metadata.A1);
+        I1 = ConvertMatrix<int>(metadata.I1);
+        Ad1 = ConvertMatrix(metadata.Ad1);
+    }
+
+    const Constants& MetaDataCudaHost::GetConstants() const
+    {
+        return m_constants;
+    }
+
+    void MetaDataCudaHost::CalcUVW(std::vector<casacore::MVuvw>& uvws)
+    {
+        this->oldUVW = uvws;
+        auto size = uvws.size();
+        uvws.clear();
+        for(int n = 0; n < size; n++)
+        {
+            auto uvw = icrar::Dot(uvws[n], dd);
+            uvws.push_back(uvw);
+        }
+    }
+
+    void MetaDataCudaHost::SetDD(const casacore::MVDirection& direction)
+    {
+        m_constants.dlm_ra = direction.get()[0] - m_constants.phase_centre_ra_rad;
+        m_constants.dlm_dec = direction.get()[1] - m_constants.phase_centre_dec_rad;
+
+        dd(0,0) = cos(m_constants.dlm_ra) * cos(m_constants.dlm_dec);
```

Comment:
`std::cos` / `std::sin` across all invocations, or `using std::cos` / `using std::sin` inside the function body.

Reply:
Better yet, see the next comment below.